File: app/src/main/java/com/tokyonth/installer/activity/SilentlyInstallActivity.java
package com.tokyonth.installer.activity;

import android.net.Uri;
import android.util.Log;

import com.tokyonth.installer.Contents;
import com.tokyonth.installer.R;
import com.tokyonth.installer.apk.APKCommander;
import com.tokyonth.installer.base.BaseActivity;
import com.tokyonth.installer.bean.ApkInfoBean;
import com.tokyonth.installer.apk.CommanderCallback;
import com.tokyonth.installer.utils.file.SPUtils;

public class SilentlyInstallActivity extends BaseActivity implements CommanderCallback {

    private APKCommander apkCommander;

    @Override
    public int setActivityView() {
        return 0;
    }

    @Override
    public void initActivity() {
        if (getIntent().getData() != null) {
            String apkSource = getIntent().getStringExtra("apkSource");
            apkCommander = new APKCommander(this, getIntent().getData(), this, apkSource);
        } else {
            showToast(getString(R.string.unable_to_install_apk));
            finish();
        }
    }

    @Override
    public void onStartParseApk(Uri uri) {
        showToast(getString(R.string.parsing));
    }

    @Override
    public void onApkParsed(ApkInfoBean apkInfo) {
        apkCommander.startInstall();
    }

    @Override
    public void onApkPreInstall(ApkInfoBean apkInfo) {
        showToast(getString(R.string.start_install, apkInfo.getApkFile().getPath()));
    }

    @Override
    public void onApkInstalled(ApkInfoBean apkInfo, int resultCode) {
        if (resultCode == 0) {
            showToast(getString(R.string.apk_installed, apkInfo.getAppName()));
            if (!apkInfo.isFakePath() && (boolean)SPUtils.getData(Contents.SP_AUTO_DEL, false)) {
                showToast(getString(R.string.apk_deleted, apkInfo.getApkFile().getName()));
            }
        } else {
            showToast(getString(R.string.install_failed, apkInfo.getAppName()));
        }
        finish();
    }

    @Override
    public void onInstallLog(ApkInfoBean apkInfo, String logText) {
        Log.e("SilentlyInstall", logText);
    }

}

File: app/src/main/java/com/tokyonth/installer/activity/SilentlyInstallActivity.kt
package com.tokyonth.installer.activity

import android.content.Intent
import android.os.Bundle
import android.view.Window
import android.view.WindowManager
import androidx.viewbinding.ViewBinding

import com.tokyonth.installer.Constants
import com.tokyonth.installer.R
import com.tokyonth.installer.install.APKCommander
import com.tokyonth.installer.data.ApkInfoEntity
import com.tokyonth.installer.data.LocalDataRepo
import com.tokyonth.installer.install.InstallCallback
import com.tokyonth.installer.install.InstallStatus
import com.tokyonth.installer.utils.AppHelper
import com.tokyonth.installer.utils.NotificationUtils
import com.tokyonth.installer.utils.ktx.string
import com.tokyonth.installer.utils.ktx.toast
import java.io.File

class SilentlyInstallActivity : BaseActivity(), InstallCallback {

    private lateinit var apkCommander: APKCommander
    private lateinit var apkInfoEntity: ApkInfoEntity

    private var installLog = ""

    override fun setBinding(): ViewBinding? = null

    override fun initView() {
        if (LocalDataRepo.instance.isFirstBoot()) {
            startActivity(Intent(this, SettingsActivity::class.java))
            finish()
        }
    }

    override fun initData() {
        val apkSource = AppHelper.reflectGetReferrer(this)!!
        if (intent.getBooleanExtra(Constants.IS_FORM_INSTALL_ACT, false)) {
            apkInfoEntity = intent.getParcelableExtra(Constants.APK_INFO)!!
            apkCommander = APKCommander(apkInfoEntity, this)
            apkCommander.startParse()
            return
        }
        intent.data.let { uri ->
            if (uri == null) {
                toast(string(R.string.unable_install_apk))
                finish()
            } else {
                if (LocalDataRepo.instance.isDefaultSilent()) {
                    apkCommander = APKCommander(uri, apkSource, this)
                    apkCommander.startParse()
                    return
                }
                Intent(this, InstallerActivity::class.java).let {
                    it.data = uri
                    it.putExtra(Constants.APK_SOURCE, apkSource)
                    startActivity(it)
                    finish()
                }
            }
        }
    }

    override fun onApkParsed(apkInfo: ApkInfoEntity) {
        if (!apkInfo.packageName.isNullOrEmpty()) {
            apkInfoEntity = apkInfo
            apkCommander.startInstall()
        } else {
            toast(string(R.string.unable_install_apk))
            finish()
        }
    }

    override fun onApkPreInstall() {
        toast(string(R.string.start_install, apkInfoEntity.appName))
    }

    override fun onApkInstalled(installStatus: InstallStatus) {
        var notificationSub = apkInfoEntity.appName!!
        val status = when (installStatus) {
            InstallStatus.SUCCESS -> {
                if (LocalDataRepo.instance.isAutoDel()) {
                    File(apkInfoEntity.filePath!!).delete()
                    notificationSub = (string(R.string.auto_del_notification, notificationSub))
                }
                string(R.string.install_successful)
            }
            InstallStatus.FAILURE -> {
                notificationSub += " ($installLog)"
                string(R.string.install_failed_msg)
            }
        }
        NotificationUtils.sendNotification(
            this,
            status,
            notificationSub,
            apkInfoEntity.getIcon()!!
        )
        finish()
    }

    override fun onInstallLog(installLog: String) {
        this.installLog = installLog
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        val attr = window.attributes.apply {
            height = 0
            width = 0
        }
        window.apply {
            attributes = attr
            setFlags(
                WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL,
                WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL
            )
            setFlags(
                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
                WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE
            )
        }
    }

}
